File: packages/repo-gc-shared/src/reporters/llm.ts
import { FINDING_KIND_LLM, SEVERITY_LLM, type Report, type Finding, FindingKind } from '../types';

function evVal(evidence: string[], key: string): string {
  const prefix = `${key}: `;
  for (const e of evidence) {
    if (e.startsWith(prefix)) return e.slice(prefix.length);
  }
  return '-';
}

function compactSummary(f: Finding): string {
  switch (f.kind) {
    case FindingKind.ContextBomb: {
      const lc = evVal(f.evidence, 'line_count');
      const tk = evVal(f.evidence, 'estimated_tokens');
      let s = `${lc}ln/${tk}tk`;
      for (const r of f.reasons) {
        if (r.includes('functions defined')) {
          const n = r.split(/\s+/)[0];
          s += ` +${n}fn`;
        } else if (r.includes('impl blocks')) {
          const n = r.split(/\s+/)[0];
          s += ` +${n}impl`;
        }
      }
      return s;
    }
    case FindingKind.CouplingHotspot: {
      const fi = evVal(f.evidence, 'fan_in');
      const fo = evVal(f.evidence, 'fan_out');
      const istab = evVal(f.evidence, 'instability');
      const pat = evVal(f.evidence, 'pattern');
      return `in=${fi} out=${fo} I=${istab} ${pat}`;
    }
    case FindingKind.DeadWeight: {
      const mp = evVal(f.evidence, 'module_path');
      const lc = evVal(f.evidence, 'line_count');
      return `${lc}ln mod=${mp}`;
    }
    case FindingKind.ReexportEntropy: {
      const pu = evVal(f.evidence, 'pub_use_count');
      const ti = evVal(f.evidence, 'total_reexported_items');
      const wc = evVal(f.evidence, 'has_wildcard');
      const w = wc === 'true' ? ' +*' : '';
      return `${ti}sym/${pu}pu${w}`;
    }
    case FindingKind.CodeDuplication: {
      const files = f.evidence.length;
      const fnName = f.evidence[0]?.split(' :: ')[1] || '?';
      return `fn:${fnName} x${files}`;
    }
    case FindingKind.UnusedImport: {
      const n = f.evidence.length;
      const names = f.evidence
        .filter(e => e.startsWith('imported but unreferenced: '))
        .map(e => e.slice('imported but unreferenced: '.length))
        .slice(0, 4);
      return `${n}: ${names.join(',')}`;
    }
    default:
      return '-';
  }
}

function compactNext(f: Finding): string {
  switch (f.kind) {
    case FindingKind.ContextBomb: {
      const lc = evVal(f.evidence, 'line_count');
      return `split <${lc}ln`;
    }
    case FindingKind.CouplingHotspot: {
      const pat = evVal(f.evidence, 'pattern');
      if (pat === 'api') return 'verify api';
      if (pat === 'orch') return 'split deps';
      return 'decouple';
    }
    case FindingKind.DeadWeight: return 'rm or re-export';
    case FindingKind.ReexportEntropy: return 'flatten re-exports';
    case FindingKind.CodeDuplication: return 'DRY: shared util';
    case FindingKind.UnusedImport: return 'rm imports';
    default: return '-';
  }
}

export function renderLlm(report: Report): string {
  const lines: string[] = [];
  const gs = report.global_score;

  const header = [
    `ver=${report.version}`,
    `fric=${gs.ai_friction_score}`,
    `waste=${gs.context_waste_score}`,
    `ent=${gs.structural_entropy_score}`,
    `ratio=${gs.context_waste_ratio.toFixed(1)}`,
    `pct=${gs.estimated_waste_pct}`,
    `files=${report.files_analyzed}`,
    `skip=${report.files_skipped}`,
    `tok=${report.total_estimated_tokens}`,
  ].join(' ');
  lines.push(header);
  if (report.errors.length > 0) {
    lines.push(`err=${report.errors.join('; ')}`);
  }

  if (report.findings.length > 0) {
    lines.push('id\tsev\tkind\tpath\ttok\tsummary\tnext');
  }

  for (const f of report.findings) {
    const sev = SEVERITY_LLM[f.severity];
    const kind = FINDING_KIND_LLM[f.kind];
    const tok = f.estimated_tokens ? `${f.estimated_tokens}tk` : '-';
    const summary = compactSummary(f);
    const next = compactNext(f);
    lines.push([f.id, sev, kind, f.path, tok, summary, next].join('\t'));
  }

  return lines.join('\n');
}

File: packages/repo-gc-python/python/repo_gc_python/reporters/llm.py
"""Token-optimized TSV reporter for LLM consumption — mirrors Rust's llm.rs."""

from ..types import Finding, FindingKind, Report


def _ev_val(evidence: list[str], key: str) -> str:
    prefix = f"{key}: "
    for e in evidence:
        if e.startswith(prefix):
            return e[len(prefix) :]
    return "-"


def _compact_summary(f: Finding) -> str:
    if f.kind == FindingKind.ContextBomb:
        lc = _ev_val(f.evidence, "line_count")
        tk = _ev_val(f.evidence, "estimated_tokens")
        s = f"{lc}ln/{tk}tk"
        for r in f.reasons:
            if "functions defined" in r:
                n = r.split()[0]
                s += f" +{n}fn"
            elif "impl blocks" in r:
                n = r.split()[0]
                s += f" +{n}impl"
        return s

    elif f.kind == FindingKind.CouplingHotspot:
        fi = _ev_val(f.evidence, "fan_in")
        fo = _ev_val(f.evidence, "fan_out")
        istab = _ev_val(f.evidence, "instability")
        pat = _ev_val(f.evidence, "pattern")
        return f"in={fi} out={fo} I={istab} {pat}"

    elif f.kind == FindingKind.DeadWeight:
        mp = _ev_val(f.evidence, "module_path")
        lc = _ev_val(f.evidence, "line_count")
        return f"{lc}ln mod={mp}"

    elif f.kind == FindingKind.ReexportEntropy:
        pu = _ev_val(f.evidence, "pub_use_count")
        ti = _ev_val(f.evidence, "total_reexported_items")
        wc = _ev_val(f.evidence, "has_wildcard")
        w = " +*" if wc == "true" else ""
        return f"{ti}sym/{pu}pu{w}"

    elif f.kind == FindingKind.CodeDuplication:
        files = len(f.evidence)
        fn_name = "?"
        if f.evidence:
            parts = f.evidence[0].split(" :: ", 1)
            if len(parts) > 1:
                fn_name = parts[1]
        return f"fn:{fn_name} x{files}"

    elif f.kind == FindingKind.UnusedImport:
        n = len(f.evidence)
        names: list[str] = []
        for e in f.evidence:
            if e.startswith("imported but unreferenced: "):
                names.append(e[len("imported but unreferenced: ") :])
        return f"{n}: {','.join(names[:4])}"

    return "-"


def _compact_next(f: Finding) -> str:
    if f.kind == FindingKind.ContextBomb:
        lc = _ev_val(f.evidence, "line_count")
        return f"split <{lc}ln"
    elif f.kind == FindingKind.CouplingHotspot:
        pat = _ev_val(f.evidence, "pattern")
        if pat == "api":
            return "verify api"
        elif pat == "orch":
            return "split deps"
        return "decouple"
    elif f.kind == FindingKind.DeadWeight:
        return "rm or re-export"
    elif f.kind == FindingKind.ReexportEntropy:
        return "flatten re-exports"
    elif f.kind == FindingKind.CodeDuplication:
        return "DRY: shared util"
    elif f.kind == FindingKind.UnusedImport:
        return "rm imports"
    return "-"


def render_llm(report: Report) -> str:
    gs = report.global_score
    header = (
        f"repo-gc\tfric={gs.ai_friction_score}\twaste={gs.context_waste_score}"
        f"\tent={gs.structural_entropy_score}\tpct={gs.estimated_waste_pct}"
        f"\trat={gs.context_waste_ratio:.1f}\tf={report.files_analyzed}"
        f"\tln={report.total_lines}\ttok={report.total_estimated_tokens // 1000}k\n"
    )

    if not report.findings:
        return header

    lines = [header, "id\tsev\tkind\tpath\ttok\tsummary\tnext\n"]

    sorted_findings = sorted(report.findings, key=lambda f: f.severity.weight, reverse=True)

    for f in sorted_findings:
        tok = str(f.estimated_tokens) if f.estimated_tokens is not None else "-"
        summary = _compact_summary(f)
        next_step = _compact_next(f)
        lines.append(
            f"{f.id}\t{f.severity.llm_label}\t{f.kind.llm_label}\t"
            f"{f.path}\t{tok}\t{summary}\t{next_step}\n"
        )

    return "".join(lines)
